packages/browser/__tests__/Session.spec.ts

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2020 Inrupt Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+// The following is required by tsyringe
+import "reflect-metadata";
+import { it, describe } from "@jest/globals";
+import { mockClientAuthentication } from "../src/__mocks__/ClientAuthentication";
+import { Session } from "../src/Session";
+import { mockStorage } from "../../core/src/storage/__mocks__/StorageUtility";
+
+describe("Session", () => {
+  describe("constructor", () => {
+    it("accepts an empty config", async () => {
+      const mySession = new Session({});
+      expect(mySession.info.isLoggedIn).toEqual(false);
+      expect(mySession.info.sessionId).toBeDefined();
+    });
+
+    it("accepts no config", async () => {
+      const mySession = new Session();
+      expect(mySession.info.isLoggedIn).toEqual(false);
+      expect(mySession.info.sessionId).toBeDefined();
+    });
+
+    it("does not generate a session id if one is provided", () => {
+      const mySession = new Session({}, "mySession");
+      expect(mySession.info.sessionId).toEqual("mySession");
+    });
+
+    it("accepts input storage", async () => {
+      const insecureStorage = mockStorage({});
+      const secureStorage = mockStorage({});
+      const mySession = new Session({
+        insecureStorage,
+        secureStorage,
+      });
+      const clearSecureStorage = jest.spyOn(secureStorage, "delete");
+      const clearInsecureStorage = jest.spyOn(insecureStorage, "delete");
+      await mySession.logout();
+      expect(clearSecureStorage).toHaveBeenCalled();
```

Comment:
I don't like these mock tests that are dependent on implementation details (i.e. that specific internal methods are being called). The name of test ("accepts input storage") seems to imply this test is just testing that the those options are 'acceptable', so what's that got to do with logging out and clearing storage? If the test is testing that the Session does in fact **_use_** the storage instances being passed into the constructor, then shouldn't it put something in that storage, and then assert that it's actually there (and that it's then 'gone' after a logout)? But that sounds like separate tests to me, which you probably have below, so I'm not sure I'd have a specific test at all for any classes constructor that it accepts certain input params, unless that constructor was performing some action on those params (i.e. what 'behaviour' is this test testing...?).

Reply:
My thinking was the following: for testing purpose, one way to use the constructor (namely passing in a ClientAuthentication) is much more convenient, so the other ways of using the constructor (e.g. here passing the `storage`) wouldn't be covered if the constructor wasn't tested independently. Here, more than testing that the specific methods are called, what I want to test is that these methods are called **on the objects passed to the constructor**, which verifies that the input parameters are properly used.

Reply:
Yeah, sure - but it's the behaviour on **_those objects passed to the constructor_** we want to test, and not the internal methods being used to enact that behaviour. So yeah, in this case I would try and store something in those mocked storage instances, then logout and then check that that storage has been cleared out (that might require some implementation knowledge, e.g. that a specific key was used to store data in that storage, but that key would be part of the documentation (i.e. our SDK uses the key 'XYZ' to store information in secure storage), so that would be Ok for this test to know about too, as it isn't a hidden internal implementation detail).